# Lab 04- Radius of The Earth

### Lillian Hanson, 09/24/2025, GPGN318A

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.constants import foot

#### Processing Calibrations
The relative gravimeters need to be calibrated to give us actual mGal values. The CG5 already accounts for this and the number it gives you is in units of mGals. For the LNR though, you must account for the calibration by hand, we have attached the (somewhat outdated) calibration sheet for our specific LNR meter. You can use this to calculate the relative gravity from the dial reading as follows:

Denote your dial reading as A.
In the supplied table (see attachment), find the nearest integer smaller than your reading from the column labeled “Counter Reading”, denote this as B.
Find the corresponding mGal value in the column labeled as “Value in mGal”, denote this as C.
Find the corresponding factor value from the column labeled as “Factor for Interval”, denote this as F.
Calculate the gravity in mGal using the formula: $g = C + F * (A - B)$.
Note: You might notice this equation as adjusting the gravity value using a linear interpolation between two known values.

In [2]:
df_lf = pd.read_csv('data/LNR_gravimeter(Sheet1).csv')
df_lf["B"] = None
df_lf["C"]= None
df_lf["F"]= None
df_lf = df_lf.rename(columns={'Digital Reading':'A'})

In [3]:
# add in values from table


# B-Values
df_lf.loc[0, 'B'] = 3100
df_lf.loc[1, 'B'] = 3100
df_lf.loc[2, 'B'] = 3100
df_lf.loc[3, 'B'] = 3100
df_lf.loc[4, 'B'] = 3100
df_lf.loc[5, 'B'] = 3100
df_lf.loc[6, 'B'] = 3100
df_lf.loc[7, 'B'] = 3100
df_lf.loc[8, 'B'] = 3100

# C-Values
df_lf.loc[0, 'C'] = 3169.70
df_lf.loc[1, 'C'] = 3169.70
df_lf.loc[2, 'C'] = 3169.70
df_lf.loc[3, 'C'] = 3169.70
df_lf.loc[4, 'C'] = 3169.70
df_lf.loc[5, 'C'] = 3169.70
df_lf.loc[6, 'C'] = 3169.70
df_lf.loc[7, 'C'] = 3169.70
df_lf.loc[8, 'C'] = 3169.70

# F-Values
df_lf.loc[0, 'F'] = 1.02481
df_lf.loc[1, 'F'] = 1.02481
df_lf.loc[2, 'F'] = 1.02481
df_lf.loc[3, 'F'] = 1.02481
df_lf.loc[4, 'F'] = 1.02481
df_lf.loc[5, 'F'] = 1.02481
df_lf.loc[6, 'F'] = 1.02481
df_lf.loc[7, 'F'] = 1.02481
df_lf.loc[8, 'F'] = 1.02481

#lol they're all the same and i put them each in manually rip

# Add row with computed G-values
df_lf["mgal"] = 3169.70 + 1.02481 * (df_lf["A"] - 3100)

df_lf

Location         A  Times            Operator     B       C        F  \
0       1.0  3144.200  13:57       Anna Williams  3100  3169.7  1.02481   
1       1.0  3144.350  14:01          Ben Koller  3100  3169.7  1.02481   
2       1.0  3144.080  14:05  Zach Shallenberger  3100  3169.7  1.02481   
3       2.0  3142.640  14:30  Sam & Tessa & Tyna  3100  3169.7  1.02481   
4       3.0  3139.242  14:45             Krishna  3100  3169.7  1.02481   
5       3.0  3139.300  14:50                 Joe  3100  3169.7  1.02481   
6       2.0  3142.830  14:59              Anna G  3100  3169.7  1.02481   
7       2.0  3142.705  15:04               Clara  3100  3169.7  1.02481   
8       1.0  3144.335  15:14                Alex  3100  3169.7  1.02481   

          mgal  
0  3214.996602  
1  3215.150323  
2  3214.873625  
3  3213.397898  
4  3209.915594  
5  3209.975033  
6  3213.592612  
7  3213.464511  
8  3215.134951

#### Measurement Averaging
Treat each instrument separately!

Average the gravity measurements taken at each visit to the stations. We visisted the stations in the following order:

- Station 1 : visit 1 
- Station 2 : visit 1
- Station 3 : visit 1
- Station 2 : visit 2
- Station 1 : visit 2
So you'll need to average all of the LNR readings at Station 1 : visit 1 together to obtain a single measurement for that location and time. Repeat with each of the visits to each station for each instrument (treat the two CG-5's separately).

At the end should have a single measurement for each of the three instruments at every visit: 5 for the LNR, 5 for CG-5 1 and 5 for CG-5 2.

#### Load in CG5 Data 

In [4]:
column_names = [
    "LINE","STATION","ALT","GRAV","SD","TILTX","TILTY","TEMP","TIDE",
    "DUR","REJ","TIME","DEC_TIME_DATE","TERRAIN","DATE"
]

df_cg51 = pd.read_csv('data/CG5-1.TXT', sep=r'\s+', skiprows=34, header=None, names=column_names)
print(df_cg51)

    LINE  STATION      ALT      GRAV     SD  TILTX  TILTY  TEMP   TIDE  DUR  \
0    1.0      1.0  29.1014  3621.271  0.426    8.9   -4.5 -1.10  0.016   30   
1    1.0      1.0  30.0779  3621.424  0.459   -5.3   -5.9 -1.11  0.023   30   
2    1.0      1.0  30.0779  3621.361  0.135   -2.9   -7.6 -1.15  0.026   30   
3    1.0      1.0  29.8338  3621.371  0.791   -4.6   -1.6 -1.18  0.030   30   
4    1.0      2.0  28.8572  3619.895  0.217   -3.0   -4.0 -1.14  0.040   30   
5    1.0      2.0  28.3689  3619.918  0.373    2.7   -0.7 -1.13  0.043   30   
6    1.0      2.0  27.8807  3619.847  0.226   -3.3   -3.5 -1.15  0.046   30   
7    1.0      3.0  27.3924  3616.459  0.242    0.4   -1.9 -1.18  0.053   30   
8    1.0      3.0  27.3924  3616.465  0.193    2.8   -9.7 -1.18  0.055   30   
9    1.0      3.0  27.3924  3616.481  0.287    1.2   -0.6 -1.19  0.058   30   
10   1.0      2.0  27.1482  3619.872  0.138   -6.0    0.6 -1.19  0.064   30   
11   1.0      2.0  26.9041  3619.902  0.161    4.1  

In [5]:
df_cg52 = pd.read_csv('data/CG5-2.TXT', sep=r'\s+', skiprows=34, header=None, names=column_names)
#print(df_cg52)
print(df_cg52.columns)

Index(['LINE', 'STATION', 'ALT', 'GRAV', 'SD', 'TILTX', 'TILTY', 'TEMP',
       'TIDE', 'DUR', 'REJ', 'TIME', 'DEC_TIME_DATE', 'TERRAIN', 'DATE'],
      dtype='object')


#### Assign variables to loaded values

In [6]:
lnr_grav = np.array(df_lf['mgal'].values)
lnr_time = np.array(df_lf['Times'].values)

times_dec = []
for time in df_lf['Times']:
    time = str(time)
    decimal_time = float(time.split(':')[0])+ float(time.split(':')[1])/60
    times_dec.append(float(decimal_time))

lnr_time = times_dec

cg5_1_grav = np.array(df_cg51['GRAV'].values)
cg5_1_time = np.array(df_cg51['DEC_TIME_DATE'].values)

cg5_2_grav = np.array(df_cg52['GRAV'].values)
cg5_2_time = np.array(df_cg52['DEC_TIME_DATE'].values)

lnr_range = np.ptp(lnr_grav)
cg51_range = np.ptp(cg5_1_grav)
cg52_range = np.ptp(cg5_2_grav)

print(f' LNR Values: {lnr_grav} mGal\n\n CG51 Values: {cg5_1_grav} mGal\n\n CG52 Values: {cg5_2_grav} mGal')
print(f' LNR Variability: {lnr_range:.2f} mGal \n CG51 Variability: {cg51_range:.2f} mGal\n CG52 Variability: {cg52_range:.2f} mgal')

 LNR Values: [3214.996602   3215.1503235  3214.8736248  3213.3978984  3209.91559402
 3209.975033   3213.5926123  3213.46451105 3215.13495135] mGal

 CG51 Values: [3621.271 3621.424 3621.361 3621.371 3619.895 3619.918 3619.847 3616.459
 3616.465 3616.481 3619.872 3619.902 3619.925 3621.43  3621.428 3621.419] mGal

 CG52 Values: [3823.572 3823.53  3823.541 3823.466 3822.116 3822.041 3822.078 3818.681
 3818.649 3818.702 3822.158 3822.149 3822.153 3823.682 3823.667 3823.658] mGal
 LNR Variability: 5.23 mGal 
 CG51 Variability: 4.97 mGal
 CG52 Variability: 5.03 mgal


### Question 1
Comment on the variability of the measurements at each visit for each instrument.

The code cell above calculates the variability for each instrument: $$ LNR:\space 5.23 \space mGal$$ $$ CG51:\space 4.97 \space mGal $$ $$ CG52:\space 5.03 \space mGal $$ 
The variability on each instrument is similar, even though the instruments measured different values. The variability could be due to error sources such as movements near the instrument during measurement. The instruments collecting different values is due to correction settings on the CG5 instruments, and likely due to human error on the LNR. 

#### Drift Correction
There are many things that could change the actual acceleration of gravity throughout the time of our survey. To correct for this, we assume that the change is linear in time and repeated the measurement at station 1 at the end of the survey (we made a loop).

Create a function that linearly interpolates the gravity values at Station 1 as a function of time:

In [7]:
def grav_interp_new(grav_1, grav_2, time_1, time_2, time):
    time = np.array(time, dtype=float)  # ensure numeric array
    slope = (grav_2 - grav_1) / (time_2 - time_1)
    new_grav = grav_1 + ( slope * (time - time_1))
    return new_grav

You will need to calculate the value of gravity at station 1 at each time for each of the instruments. Then subtract that value from every measurement. If done correctly the repeated values of station 1 at the start and end should be the same and equal to 0.

In [8]:
lnr_drift_corr = lnr_grav + grav_interp_new(lnr_grav[0], lnr_grav[-1], lnr_time[0], lnr_time[-1], lnr_time)
#assert lnr_drift_corr[0] == lnr_drift_corr[-1] == 0

cg51_drift_corr = cg5_1_grav + grav_interp_new(cg5_1_grav[0], cg5_1_grav[-1], cg5_1_time[0], cg5_1_time[-1], cg5_1_time)
cg52_drift_corr = cg5_2_grav + grav_interp_new(cg5_2_grav[0], cg5_2_grav[-1], cg5_2_time[0], cg5_2_time[-1], cg5_2_time)

cg51_drift_corr = pd.Series(cg51_drift_corr, name='DRIFT_CORR')
df_cg51 = pd.concat([df_cg51, cg51_drift_corr], axis=1)

cg52_drift_corr = pd.Series(cg52_drift_corr, name='DRIFT_CORR')
df_cg52 = pd.concat([df_cg52, cg52_drift_corr], axis=1)

lnr_drift_corr = pd.Series(lnr_drift_corr, name='DRIFT_CORR')
df_lnr = pd.concat([df_lf, lnr_drift_corr], axis=1)

### Question 2
We repeated station 2 twice as well, how close are the values at this station after drift corrections?

The values at station 2 at each visit are within 3 mGal of each other on each instrument. 

#### Relative Gravity at each station
For each instrument you should now have their drift corrected relative gravity measurements. Since you have subtract the value of gravity of Station 1 from every Station, your measurements are relative to Station 1.

Next, for each instrument average any values at repeated stations to obtain a single measurement at each station (independant of time).

In [10]:
cg51_avg = np.array(df_cg51.groupby('STATION')['DRIFT_CORR'].mean())
print(cg51_avg)

cg52_avg = np.array(df_cg52.groupby('STATION')['DRIFT_CORR'].mean())
print(cg52_avg)

lnr_avg =  np.array(df_lnr.groupby('Location')['DRIFT_CORR'].mean())
print(lnr_avg)

[7242.72825262 7241.25239266 7237.8286968 ]
[7647.20131325 7645.73909448 7642.30173946 7642.30092822]
[6430.07545498 6428.57863347 6425.03265112]


In [11]:
# These arrays should end up with your final relative
# values at each of the three stations for each instrument
#lnr_rel = np.array([lnr_1_avg, lnr_2_avg, lnr_3_avg])
cg51_rel = np.array([cg51_avg[0], cg51_avg[1], cg51_avg[2]])
cg52_rel = np.array([cg52_avg[0], cg52_avg[1], cg52_avg[2]])
lnr_rel = np.array([lnr_avg[0], lnr_avg[1], lnr_avg[2]])
print(f'{cg51_rel}\n{cg52_rel}\n{lnr_rel}')

[7242.72825262 7241.25239266 7237.8286968 ]
[7647.20131325 7645.73909448 7642.30173946]
[6430.07545498 6428.57863347 6425.03265112]


#### Relative to Absolute
Both the CG5 and the LNR are relative gravimeters, meaning they do not measure the full gravity value, instead they measure differences in gravity between two stations. The units of their data are in mGal (after calibration), the difference between 1 mGal and another is accurate, but a single value itself is meaningless.

Ex. Imagine someone telling you it was 1.4233 degrees celsius warmer inside than it was outside. They obtained this number by using an instrument that read out 331$^\circ$C and 332.4233$^\circ$C. The value itself doesn’t mean anything, but the difference between them is accurate. If the true temperature inside was 24$^\circ$C, then you could use this hypothetical relative temperature sensor to determine that the temperature outside was actually 24.4233$^\circ$C.

Because we know for certain the gravity value at one of the stations, we can use its value, and the relative differences from the gravimeters, to determine the absolute gravity at each of the other stations.

After drift corrections your data are explicitly the differences between each visit and station 1 ($\Delta S_i$). Use these values to calculate the absolute gravity at each station.
$$\cdot g_{abs, S_2}= \Delta (S_2)+g_{abs, S_1}$$


In [12]:
# calculate absolute grav values

lnr_1_abs = lnr_rel[0]
lnr_2_abs = lnr_rel[1] + lnr_rel[0]
lnr_3_abs = lnr_rel[2] + lnr_rel[0]

cg51_1_abs = cg51_rel[0]
cg51_2_abs = cg51_rel[1] + cg51_rel[0]
cg51_3_abs = cg51_rel[2] + cg51_rel[0]

cg52_1_abs = cg52_rel[0]
cg52_2_abs = cg52_rel[1] + cg52_rel[0]
cg52_3_abs = cg52_rel[2] + cg52_rel[0]

### Question 3
For each instrument calculate an approximate gradient between stations 2 and 3.
$$\partial_zg \approx \frac{gs_3-gs_2}{\Delta H}$$ How close is this to the theoretical free air gradient of -0.3086 mGal/m? Comment on any sources of error you might have encountered.

 $\Delta H$ = 40 m

In [13]:
# Convert H value to meters
H_ft = 40
H = 40 * foot

# calculate and print gradients and differences
grad_cg51 = (cg51_3_abs - cg51_2_abs) / H
grad_cg52 = (cg52_3_abs - cg52_2_abs) / H
grad_lnr = (lnr_3_abs - lnr_2_abs) / H

print(f'LNR Gradient: {grad_lnr:.4f} mGal/m\nCG51 Gradient: {grad_cg51:.4f} mGal/m\nCG52 Gradient: {grad_cg52:.4f} mGal/m')
print('-----------------')
print(f'LNR difference: {grad_lnr-(-0.3086):.4f} mGal/m\nCG51 difference: {grad_cg51-(-0.3086):.4f} mGal/m\nCG52 difference: {grad_cg52-(-0.3086):.4f} mGal/m')

LNR Gradient: -0.2908 mGal/m
CG51 Gradient: -0.2808 mGal/m
CG52 Gradient: -0.2819 mGal/m
-----------------
LNR difference: 0.0178 mGal/m
CG51 difference: 0.0278 mGal/m
CG52 difference: 0.0267 mGal/m


These values are fairly close to the theoretical free air gradient, $-0.3086 \space mGal/m$, with all calculated values within $0.027 \space mGal/m$ of the free air gradient. Sources of error could include:
- Vibrations near the instrument( people walking, cars moving, etc.)
- Human error in adjusting the LNR gravimeter
- Inaccurate record of time or height

## Determining the radius of the Earth
With the absolute gravity values at each location we can now find the radius of the Earth, without needing to know the mass (or the gravitational constant)! We are going to use a first order Taylor series expansion of Newton’s Universal Law of Gravitation:

### Question 4
Find an expression for the radius of the Earth at Station 2.

First, find the derivative of Newton's law with respect to $R$
$$\frac{\partial g_{abs}}{\partial R} = ...$$
Then simplify this using $g_{abs}(R)$ itself. i.e.
$$\frac{\partial g_{abs}}{\partial R}=g_{abs}(R)...$$
Use Newton's law in the Taylor series with $f(a)=g_{abs}(S_2)$ and $f(x)=g_{abs}(S_3)$ including only the first order in $x$ to show that
$$g_{abs}(S_3) \approx g_{abs}(S_2)-2\frac{g(S_2)}{R_2}\Delta H$$
Finally solve for $R_2$ to obtain an expression for the radius of the Earth at Station 2. 

In [14]:
# to solve for R2- LNR:
g2_lnr = lnr_2_abs
g3_lnr = lnr_3_abs
h = H # converted value from above

r2_lnr = (2 * g2_lnr * h) / (g2_lnr - g3_lnr)

print(f'g2 = {g2_lnr:.4f}\ng3 = {g3_lnr:.4f}\nLNR Radius at S2 = {r2_lnr / 100:.4f} km')

g2 = 12858.6541
g3 = 12855.1081
LNR Radius at S2 = 884.2272 km


##### Radius of the Earth measured by the LNR instrument at Station 2: 884.2272 km

### Question 5
Use your equation to obtain estimates of the Earth for each instrument. Comment on how close they were to each other and how close they were to the distance to the center of the Earth in Golden. 

In [15]:
# to solve for R2-cg51:
g2_cg1 = cg51_2_abs
g3_cg1 = cg51_3_abs

r2_cg51 = (2 * g2_cg1 * h) / (g2_cg1 - g3_cg1)

In [16]:
# to solve for R2-cg52:
g2_cg2 = cg52_2_abs
g3_cg2 = cg52_3_abs

r2_cg52 = (2 * g2_cg2 * h) / (g2_cg2 - g3_cg2)

In [17]:
print(f'LNR Radius at S2 = {r2_lnr / 100:.4f} km')
print(f'CG51 Radius at S2 = {r2_cg51 / 100:.4f} km')
print(f'CG52 Radius at S2 = {r2_cg52 / 100:.4f} km')

LNR Radius at S2 = 884.2272 km
CG51 Radius at S2 = 1031.5676 km
CG52 Radius at S2 = 1084.8547 km


#### Results:

My calculated values for the radius of the Earth are printed above. I am aware that this is not a realistic value for the radius of the Earth. This error definitely occured with the way I processed the data. My educated guess is that the error/bug specifically lies in my drift correction code, but I was not able to resolve this issue at this time. 

### Deliverables:
- Describe the steps necessary to take a reading with each gravimeter
- Answer the questions throughout the report (5 total)
- Present the results of your attempts to measure the radius of the Earth.
- Attatch your field notebook to your report.

### Steps for each Gravimeter:

#### LaCoste and Romberg (LNR) Model G gravity meter
Getting measurements:
1) Level the gravimeter by adjusting the screws and aligning the leveling bubbles.
2) Unlock the spring
3) Turn the light on
4) Looking into the eyepiece, use the adjustment dial to get the scale to read 2.4 (ensure the reading is stable)
5) Record the reading next to the adjustment dial
7) Convert measurement values according to manual guide

#### Scintrex CG-5 gravimeter
Getting measurements:
1) Ensure the station label is correct
2) Level the instrument until the black axis line align in the center circle (you should get a smiley face)
3) Press measure and move away from the instrument
4) The blue light indicates it is in the process of taking measurements- reduce movement while this is happening
5) Save the measurement